test/pylib/log_browsing.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2023-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+import logging
+from typing import Optional
+import pytest
+import os
+import re
+
+
+logger = logging.getLogger(__name__)
+
+class ScyllaLogFile():
+    """
+    Class for browsing a Scylla log file.
+    Based on scylla-ccm implementation of log browsing.
+    """
+    def __init__(self, thread_pool: ThreadPoolExecutor, logfile_path: str):
+        self.thread_pool = thread_pool # used for asynchronous IO operations
+        self.file = logfile_path
+        if not os.path.isfile(self.file):
+            pytest.fail("Log file {} does not exist".format(self.file))
+
+    async def _run_in_executor(self, func, *args, loop=None):
+        if loop is None:
+            loop = asyncio.get_running_loop()
+        return await loop.run_in_executor(self.thread_pool, func, *args)
+
+    async def mark(self) -> int:
+        """
+        Returns "a mark" to the current position of this node Scylla log.
+        This is for use with the from_mark parameter of watch_log_for method,
+        allowing to watch the log from the position when this method was called.
+        """
+        with open(self.file, 'r') as f:
+            await self._run_in_executor(f.seek, 0, os.SEEK_END)
+            return await self._run_in_executor(f.tell)
+
+    async def wait_for(self, pattern: str | re.Pattern, from_mark: Optional[int] = None, timeout: int = 600) -> None:
+        """
+        wait_for() checks if the log contains the given message.
+        Because it may take time for the log message to be flushed, and sometimes
+        we may want to look for messages about various delayed events, this
+        function doesn't give up when it reaches the end of file, and rather
+        retries until a given timeout. The timeout may be long, because successful
+        tests will not wait for it. Note, however, that long timeouts will make
+        xfailing tests slow.
+        The timeout is in seconds.
+        If from_mark is given, the log is searched from that position, otherwise
+        from the beginning.
+        """
+        prog = re.compile(pattern)
+        loop = asyncio.get_running_loop()
+
+        with open(self.file, 'r') as f:
+            if from_mark is not None:
+                await self._run_in_executor(f.seek, from_mark, loop=loop)
+
+            async with asyncio.timeout(timeout):
+                logger.debug("Waiting for log message: %s", pattern)
+                while True:
+                    line = await self._run_in_executor(f.readline, loop=loop)
+                    if line and prog.search(line):
+                        logger.debug("Found log message: %s", line)
+                        return
```

Comment:
If readline reached EOF, do you try it immediately, again and again, until it starts returning data? Did you test that this actually works (it might not work if Python's stdio library caches the fact it reached EOF!!).
Also, you should probably add a short sleep between retries - no point in doing this loop a million times a second :-(

Reply:
I wrote a short test for this and found that:
1. readline() does work after it returned EOF once (it's not cached). So your code seems correct even if not obvious.
2. While waiting, it does a million (!) iterations per second, which is completely excessive. So a sleep, even a very short one (e.g., 0.01 seconds) is probably better than nothing.